### DISCUSSION TITLE
Should we add a benchmark in test dir?

### DISCUSSION THREAD

**Discussion by andreivladbrg** [sablier-labs/evm-monorepo#875] [Ideas]
As per a private discussion with @smol-ninja, we agreed that it would be a good idea to add a benchmark for gas usage to all relevant functions: `createWithDurations`, `createWithTimestamps`, `withdraw`, `cancel`, etc.

Now, the question is: where would be the best place to include them? I can think of two options:
- we add a new directory called benchmarks, where we can have a single contract and add a test for each targeted function. This way, we have them grouped together.
- we add, at the end of each concrete test, one more function for gas cost ([example of file](https://github.com/sablier-labs/v2-core/blob/staging/test/integration/concrete/lockup-linear/create-with-timestamps/createWithTimestamps.t.sol)). With this option, we wouldn't need to implement a new contract.

Note: It would be a good idea to name the benchmark tests in a consistent manner, like `testGas_FunctionName`, so that we can run a specific CLI command to obtain them: `forge t --mt "testGas"`

@PaulRBerg, what do you say?

**Comment by PaulRBerg** [comment]
Good idea.

Let's go with the first option and create a dedicated `benchmarks` directory. This is a standard software development practice.

Adding functions at the end of each existing contract would clog up the tests, and it would make it difficult to integrate with Bulloak in the future.

**Reply by andreivladbrg** [reply]
> it would make it difficult to integrate with Bulloak in the future.

good point, agree

**Comment by smol-ninja** [comment]
This is also be a pre-requirement for https://github.com/sablier-labs/v2-core/issues/872 since https://github.com/sablier-labs/v2-core/issues/872 requires gas calculations in order to compute max count of segments/tranches.

**Comment by andreivladbrg** [comment]
Created the issue: https://github.com/sablier-labs/v2-core/issues/876
